grove/alpha/simon/simon.py

```diff
@@ -14,440 +14,316 @@
-                     :math:`\\vert y \\rangle`.
-    :param str gate_name: Optional parameter specifying the name of
-                          the gate that will represent unitary_funct
-    :return: A program that performs the above unitary transformation.
-    :rtype: Program
-    """
-    assert is_unitary(unitary_funct), "Function must be unitary."
-    p = pq.Program()
-
-    inverse_gate_name = gate_name + '-INV'
-    scratch_bit = p.alloc()
-    bits_for_funct = [scratch_bit] + qubits
-
-    p.defgate(gate_name, unitary_funct)
-    p.defgate(inverse_gate_name, np.linalg.inv(unitary_funct))
-    p.inst(tuple([gate_name] + bits_for_funct))
-    p.inst(list(map(lambda qs: CNOT(qs[0], qs[1]), zip(qubits, ancillas))))
-    p.inst(tuple([inverse_gate_name] + bits_for_funct))
-
-    p.free(scratch_bit)
-    return p
+from pyquil.gates import H
+import grove.alpha.simon.utils as u
+from collections import defaultdict
 
 
-def simon(oracle, qubits):
+def create_periodic_1to1_bitmap(mask):
     """
-    Implementation of the quantum portion of Simon's Algorithm.
-
-    Given a list of input qubits,
-    all initially in the :math:`\\vert 0\\rangle` state,
-    create a program that applies the Hadamard-Walsh transform the qubits
-    before and after going through the oracle.
-
-    :param Program oracle: Program representing unitary application of function
-    :param list(int) qubits: List of qubits that enter as the input
-                        :math:`\\vert x \\rangle`.
-    :return: A program corresponding to the desired instance of
-             Simon's Algorithm.
-    :rtype: Program
+    A helper to create a bit map function for a given mask. E.g. for a mask :math:`m = 10` the
+    return would is a dictionary:
+
+    >>> create_periodic_1to1_bitmap('10')
+    ... {
+    ...     '00': '10',
+    ...     '01': '11',
+    ...     '10': '00',
+    ...     '11': '01'
+    ... }
+
+    :param mask: binary mask as a string of 0's and 1's
+    :return: dictionary containing a mapping of all possible bit strings of the same size a the mask
+    and their mapped bit-string value
+
+    :rtype: Dict[String, String]
     """
-    p = pq.Program()
+    n_bits = len(mask)
+    form_string = "{0:0" + str(n_bits) + "b}"
+    dct = {}
+    for idx in range(2**n_bits):
+        bit_string = form_string.format(idx)
+        dct[bit_string] = u.bitwise_xor(bit_string, mask)
+    return dct
 
-    # Apply Hadamard, Unitary function, and Hadamard again
-    p.inst(list(map(H, qubits)))
-    p += oracle
-    p.inst(list(map(H, qubits)))
-    return p
 
-
-def is_unitary(matrix):
+def create_valid_2to1_bitmap(mask):
     """
-    A helper function that checks if a matrix is unitary.
-
-    :param 2darray matrix: a matrix to test unitarity of
-    :return: true if and only if matrix is unitary
-    :rtype: bool
+    A helper to create a valid 2-to-1 function with a given mask. This can be used to create a valid
```

Comment:
"A helper to create a 2-to-1 binary function that is invariant with respect to the application of a specified XOR bitmask.  This property must be satisfied if a 2-to-1 function is to be used in Simon's algorithm."

Reply:
done!